src/lsm/table_data_iterator.zig

```diff
@@ -102,12 +118,35 @@ pub fn TableDataIteratorType(comptime Storage: type) type {
         fn on_read(read: *Grid.Read, block: Grid.BlockPtrConst) void {
             const it = @fieldParentPtr(TableDataIterator, "read", read);
             assert(it.callback == .read);
+            assert(it.context.addresses.len == it.context.checksums.len);
 
             const callback = it.callback.read;
             it.callback = .none;
-            it.context.addresses = it.context.addresses[1..];
-            it.context.checksums = it.context.checksums[1..];
 
+            switch (it.context.direction) {
+                .ascending => {
+                    if (constants.verify) {
+                        const header = schema.header_from_block(block);
+                        assert(header.op == it.context.addresses[0]);
+                        assert(header.checksum == it.context.checksums[0]);
+                    }
+
+                    it.context.addresses = it.context.addresses[1..];
+                    it.context.checksums = it.context.checksums[1..];
+                },
+                .descending => {
+                    if (constants.verify) {
+                        const header = schema.header_from_block(block);
+                        assert(header.op == it.context.addresses[it.context.addresses.len - 1]);
+                        assert(header.checksum == it.context.checksums[it.context.checksums.len - 1]);
+                    }
+
+                    it.context.addresses = it.context.addresses[0 .. it.context.addresses.len - 1];
+                    it.context.checksums = it.context.checksums[0 .. it.context.checksums.len - 1];
+                },
+            }
```

Comment:
Since we iterate in either direction now, it might be simpler to add a `Context.index: usize` field and increment/decrement that instead of re-slicing `addresses`/`checksums`.

Reply:
If you don't mind, I'd rather keep the slicing logic, it's easy to understand and prevents off-by-one bugs.